### ISSUE TITLE
Porting serenity's command framework example

### ISSUE THREAD

**Issue by s0lst1ce** [kangalioo/poise#10]
The serenity repo has a very [extensive example](https://github.com/serenity-rs/serenity/tree/current/examples/e05_command_framework) for their command framework. The later not only makes use of many of the framework's features but also gives many hints as to what is what and how to use the features.

I know `poise` is still far from done but I'm thinking trying to port this example may help drive it forward and improve user adoption. This would also alleviate the sometimes lacking documentation.

**Comment by kangalioo** [comment]
I used my free period at school to start this :) [WIP](https://pastebin.com/qyKULZu1)

**Comment by kangalioo** [comment]
Tl;dr: I'm torn on the convenience<->simplicity trade-off

The serenity framework example accesses the shard manager from the ClientBuilder and actively writes it into the user data. In poise, both ClientBuilder configuration and user data creation are handled via callbacks from within poise, so it's a bit tricky to pass data between ClientBuilder and user data.

What I had in mind was to store the shard manager in the framework itself and make it available with a `shard_manager()` method or similar. User code could then just write `ctx.framework().shard_manager()` without any setup needed.

That plan would introduce even more magic into the framework though, which is not great. I'm still unsure about the decision to handle all framework setup by a single builder ([example](https://github.com/kangalioo/poise/commit/9828d7b19cfaea93647b1fb1cdd8f691739de3ad#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5)). It made things more streamlined but at the expense of obscuring what the framework is actually doing under the hood... Maybe the user_data_setup callback should also be removed in favor of providing a fully initialized Data struct right away. That way, you won't have access to the Ready event data, but maybe that's not needed anyways... If both the ClientBuilder and user data setup callbacks were gone in favor of direct argument passing, the serenity framework approach could be used.

Sorry for my rambling, don't feel obligated to respond to all of this haha

I'll let it rest some more, maybe a good idea will emerge

**Comment by s0lst1ce** [comment]
I don't know nearly enough about poise or serenity to assess the technical challenges. However I do believe that a framework should allow the user to access the underlying primitives (or models).
Currently poise is far from feature-complete. And even if it was there would still always be edge-cases where what's provided by the framework wouldn't suffice.
Of course the goal of the framework is still to ease the development process. Hence I think the framework should allow users to access the underlying models in a usable way (ie: accessing them with at most a method and not requiring them to keep the framework updated and what they do with the primitives and data). As long as this is possible I think the focus should only be on making the framework ergonomic to use as if it wasn't used alongside anything else.
I think this would fit what you were proposing with the `.shard_manager()` method?

What's your designing philosophy for poise?

**Comment by kangalioo** [comment]
In the end I decided on `.shard_manager()`.

After making it compile, the ported example is finally pushed to the repo :) https://github.com/kangalioo/poise/blob/master/examples/serenity_example_port/main.rs

Some parts are commented out, since poise does not have support for those features
